### PR TITLE
Authn: JWT client

### PR DESCRIPTION
**Pull request by Jguer**
**What is this feature?**

- Implement JWT client for AuthN
- fixup oauth devenv instructs for keycloak 19

- Get user by login returned an empty user that could be mistaken for a valid user. Add check for that and return nil for safety

### CODE REVIEW

File: pkg/services/authn/clients/jwt.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	}
+
+	sub, _ := claims["sub"].(string)
+	if sub == "" {
+		s.log.Warn("Got a JWT without the mandatory 'sub' claim", "error", err)
+		return nil, ErrJWTMissingClaim.Errorf("missing mandatory 'sub' claim in JWT")
+	}
+
+	id := &authn.Identity{
+		AuthModule: "jwt",
```
Comment (mgyongyosi) [inline pkg/services/authn/clients/jwt.go:65]: I would introduce constants for the names of the authmodules, wdyt?

Comment (Jguer) [reply pkg/services/authn/clients/jwt.go]: done, nice catch as we already have a bunch of them made into constants

File: pkg/services/authn/clients/jwt.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	"strings"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/auth"
+	"github.com/grafana/grafana/pkg/services/authn"
+	"github.com/grafana/grafana/pkg/services/login"
+	"github.com/grafana/grafana/pkg/services/org"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util/errutil"
+	"github.com/jmespath/go-jmespath"
```
Comment (kalleep) [inline pkg/services/authn/clients/jwt.go:16]: Nit move non grafana package to its own import group

File: pkg/services/authn/clients/jwt.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+		jwtToken = r.HTTPRequest.URL.Query().Get("auth_token")
+	}
+
+	// Strip the 'Bearer' prefix if it exists.
+	jwtToken = strings.TrimPrefix(jwtToken, "Bearer ")
+
+	claims, err := s.jwtService.Verify(ctx, jwtToken)
+	if err != nil {
+		s.log.Debug("Failed to verify JWT", "error", err)
+		return nil, ErrJWTInvalid.Errorf("failed to verify JWT")
```
Comment (kalleep) [inline pkg/services/authn/clients/jwt.go:56]: Should we wrap error from jwt service?

**Review by kalleep — APPROVED** [review]
Looks good, left some nits

File: pkg/services/authn/clients/jwt.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+	if jwtToken == "" && s.cfg.JWTAuthURLLogin {
+		jwtToken = r.HTTPRequest.URL.Query().Get("auth_token")
+	}
+
+	if jwtToken == "" {
+		return false
+	}
+
+	// Strip the 'Bearer' prefix if it exists.
+	jwtToken = strings.TrimPrefix(jwtToken, "Bearer ")
```
Comment (mgyongyosi) [inline pkg/services/authn/clients/jwt.go:144]: nit: These are almost identical to https://github.com/grafana/grafana/pull/61157/files#diff-da754e78d43b1b454b09a7ee9163959b631d7622cbe7a9f495fc0acca0032ddbR45-R51. So the `jwtToken` extraction logic could be extracted to a function.

File: pkg/services/authn/clients/jwt.go (hunk 3 above)
Comment (Jguer) [reply pkg/services/authn/clients/jwt.go:56]: The previous implementation didn't do it, I'm not afraid a lot of information is being given to an attacker, we could change it

**Review by mgyongyosi — APPROVED** [review]

**Pull request merged**